extensions/amp-ad-network-mgid-impl/0.1/amp-ad-network-mgid-impl.js

```diff
@@ -0,0 +1,360 @@
+import {
+  CONSENT_POLICY_STATE,
+  CONSENT_STRING_TYPE,
+} from '#core/constants/consent-state';
+import {createElementWithAttributes, removeElement} from '#core/dom';
+
+import {Services} from '#service';
+
+import {user} from '#utils/log';
+
+import {AmpA4A} from '../../amp-a4a/0.1/amp-a4a';
+
+/** @const {string} */
+const TAG = 'amp-ad-network-mgid-impl';
+
+const BASE_URL_ = 'https://servicer.mgid.com/';
+const PV_URL_ = 'https://c.mgid.com/';
+
+export class AmpAdNetworkMgidImpl extends AmpA4A {
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.ampAnalyticsElement_ = null;
+
+    /** @private */
+    this.mgidMetadata_ = {
+      'muidn': '',
+      'pvid': '',
+    };
+  }
+
+  /** @override */
+  tearDownSlot() {
+    super.tearDownSlot();
+    if (this.ampAnalyticsElement_) {
+      removeElement(this.ampAnalyticsElement_);
+      this.ampAnalyticsElement_ = null;
+    }
+  }
+
+  /** @override */
+  isValidElement() {
+    const id = this.element.getAttribute('data-widget');
+    if (!id || parseInt(id, 10) != id) {
+      user().warn(TAG, 'Undefined or non-numeric data-widget param!');
+      return false;
+    }
+    return true;
+  }
+
+  /** @override */
+  getAdUrl(consentTuple, opt_rtcResponsesPromise) {
+    let adUrlParams = [];
+
+    const consentParams = this.getConsents_(consentTuple);
+    if (consentParams.length !== 0) {
+      adUrlParams = adUrlParams.concat(consentParams);
+    }
+
+    const widget = this.element.getAttribute('data-widget');
+
+    let servicerUrl = BASE_URL_ + widget + '/' + this.getPageParam_();
+    let pvUrl = PV_URL_ + 'pv/';
+
+    adUrlParams = adUrlParams.concat(this.getNetworkInfoParams_());
+    adUrlParams.push(this.getCacheBusterParam_());
+    adUrlParams.push(this.getDevicePixelRatioParam_());
+    adUrlParams.push(this.getCxurlParam_());
+    adUrlParams.push(this.getPrParam_());
+    adUrlParams.push(this.getLuParam_());
+    adUrlParams.push(this.getSessionIdParam_());
+    adUrlParams.push(this.getPvidParam_());
+    if (localStorage.mgMuidn) {
+      adUrlParams.push('muid=' + localStorage.mgMuidn);
+    }
+    adUrlParams.push('implVersion=15');
+
+    return Promise.allSettled(adUrlParams).then((params) => {
```

Comment:
I don't think `allSettled` is available in all our supported browsers. You may need to do a `map` then `Promise.all`

Reply:
Ok, I can fix it. But can you clarify which browsers are we talking about? Here are browsers supported by AMP - https://amp.dev/support/faq/supported-browsers And as I see on caniuse, support of Promise.all and Promise.allSettled is pretty much the same - https://caniuse.com/?search=Promise.all and https://caniuse.com/?search=Promise.allSettled

Reply:
You are right, this seems ok.